README.md
Arc
=====

Arc Web Application Framework
 - Written by Craig Longford (DeltaWolf7)
 - Email: deltawolf7@gmail.com
 - WWW: www.deltasblog.co.uk

Default Login
=============

Email: admin@server.local

Password: password


What is Arc
===========

Arc is a framework designed to facilitate the rapid development of web applications. 
This is achieved through the use of self loading modules.

Based on the BootStrap design framework, JQuery, Medoo, Font Awesome and SummerNote.

Arc Features
- User management system
- Group based permission system
- Dynamic page creation
- SEO Friendly URL construction throughout
- Simplified AJAX commands for rapid development
- Support for CSS based themes
- Intelligent 404, 403 and 419 handling
- Module system and rapid development structure
- Easy to learn, hard to forget

CHANGELOG
Bootstrap                   3.3.2               http://getbootstrap.com/
Boostrap Datetimepicker     4.0.0               http://eonasdan.github.io/bootstrap-datetimepicker/
JQuery                      2.1.3               http://jquery.com/download/
Font Awesome                4.3.0               http://fortawesome.github.io/Font-Awesome/
Summernote                  0.6.1               http://summernote.org/
Bootstrap Filestyle         1.1.2               http://markusslima.github.io/bootstrap-filestyle/
Medoo                       0.9.8               http://medoo.in/
Moment.js                   2.9.0               http://momentjs.com/
PHPMailer                   5.2.8               https://github.com/Synchro/PHPMailer



1.0.0.0
--------
/ Updated all 3rd party libraries and scripts.
+ Rewrote the entire Arc framework.
+ Added support for multi group users.
/ Changed the interface style throughout.
+ Added customer MVC style system for modules and themes.
+ Added module initialisation system.
+ Added support for desktop systems such as PHPDesktop and Sqlite.

0.0.0.41
--------
/ Changed the way dispatch for ajax is done.
/ Changed theme controllers.
/ Changed modules controllers.
+ Added per page module controllers.
+ Added per module controller.
+ Added fullname to users.
+ Favicon now set by themes not system.
/ Cleaned config.

0.0.0.40
--------
/ Cleared out namespaces, change mind for now about this.
/ Changed the way paths are detected.
+ Added get footer method.
/ Shifted javascript to footer for faster page loading.
/ Updated all modules and templates to suite changes.
- Removed unused settings from config.
/ Finished the blog module.
/ Corrected issue in askquestions causing wrong answer.
/ Fixed issue with menu building messing up some paths.
/ Changed how the autoload works.

0.0.0.37
--------
+ Added check to mail sender.
/ Changed javascript function order in some modules.
+ Added getAll to SystemSettings class.

0.0.0.36
--------
+ Added PHPMailer
/ General updates and corrections.
+ Added Arc sendmail function.
+ Added toArray function to SystemSetting class.
/ Changed the way auto load works, now detects modules.
+ Added correct version numbers to modules.
+ Added getHost function.
/ Changed the login module.
+ Added forgot password feature.
/ Fixed details not updating when changed.
- Removed override from default template.

0.0.0.35
--------
+ Added url and referer to logs.
/ Changed the way logging now works to cover all actions.
+ Added a couple of methods to the user class to prevent the admin being removed/changed on the demo.
+ Updated all the sql files in the project to be current.

0.0.0.34
--------
/ Changed included date picker, for a better datetime picker.
+ Added Moment.js

0.0.0.33
--------
/ Corrected wrong path for jquery.

0.0.0.32
--------
/ Corrected bug preventing administrative access.

0.0.0.31
--------
+ Added controller system.
/ Changed the ways menus are now added.
+ Added favicon the theme controller.
- System no longer tries to theme navbar.
+ Added view override support for styling modules.

0.0.0.30
--------
/ Changed structure of Arc to better support themes and allow manipulation.
/ Changed the way headers are built to allow custom additions.
/ Moved all functions and features to one file for easy management and inclusion. 
+ Added default theme.
+ Added two other modules used in other projects. (Can be deleted, if not required).
/ Updated all modules to meet new layout.
- Removed auto built headers.

0.0.0.29
--------
/ Optimised some core functions.

0.0.0.28
--------
+ Added powered by function to show arc version.
/ Changed the location of the bootstrap file.
/ Updated modules to work with new structure.

0.0.0.27
--------
+ Added support for modules to edit page title, meta description and meta tags.
+ Added canonical url support.
+ Added og meta tags to blog module header.
/ Changed a tags to output double quotes instead of single.
- Removed unneeded tabs in menu builder.

0.0.0.26
--------
/ Cleaned up some code and added comments.
/ Updated sql script.

0.0.0.25
--------
- Removed old themes.
+ Added new theme system to replace old theme system.
/ Changed the layout of Arc to have less in the root.
+ Added new system settings class and table for system related settings.
- Removed theme options from user details and administration module.
! Note: index.php in the root no longer requires editing to change looks and feel.

0.0.0.24
--------
+ Added links to installer for modules.

0.0.0.23
--------
+ Added max length to forms to prevent truncating.
+ Added function to edit permissions to pages administration.

0.0.0.22
--------
/ Changed the way the title is displayed on pages. (Not shown if empty)
* Fixed the pages administration module not saving.
+ Added remove page to page module.
* Fixed pages not loading.
* Fixed bug in installer that prevented modules from being deleted.
+ Added module details to installer.

0.0.0.21
--------
+ Added support for setting default theme from config.
* Fixed missing closing div tag on user management.
/ Changed the way menus are built.
+ Added support for multi menu items from one module.
+ Added support for specifying url in menu item.
- Removed requires login from menu.

0.0.0.20
--------
* Updated Medoo to version 0.9.6.2
+ Added html editor to page editor module.

0.0.0.19
--------
* Further improved the UI of the in-built modules.
* Improved registration module.

0.0.0.18
--------
+ Added error handling to AJAX library.

0.0.0.17
--------
* Fixed small typo in error module.

0.0.0.16
--------
+ Added support for group descriptions.
+ Added function to remove custom groups from permissions.
+ Added support for Bootstrap Filestyle.
/ Changed style of the installer module to match.
/ Changed the way module details are loaded. (Corrects modules with no details)
+ Added support to modules to protect from removal.
/ Improved installer layout and information.

functions.php
<?php

/*
 * The MIT License
 *
 * Copyright 2014 Craig Longford.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * Application functions
 *
 * @author Craig Longford
 */
// get database object
function arcDatabase() {
    return $GLOBALS["arc_db"];
}

// get www path or filesystem path if true
function arcGetPath($filesystem = false) {
    if ($filesystem) {
        return $_SERVER["DOCUMENT_ROOT"] . ARCFS;
    }
    return ARCWWW;
}

// url data
$arc_url_data = null;

// get url data
function arcGetURLData($name = null) {
    if ($name != null) {
        if (isset($GLOBALS["arc_url_data"][$name])) {
            return $GLOBALS["arc_url_data"][$name];
        }
        return null;
    }
    return $GLOBALS["arc_url_data"];
}

// function to split url
function arcSplitURL() {
    if (isset($_REQUEST["url"])) {
        $url = explode("/", $_REQUEST["url"]);
        $count = 0;
        foreach ($url as $item) {
            if ($count == 0) {
                $GLOBALS["arc_url_data"]["module"] = $item;
            } else {
                $GLOBALS["arc_url_data"]["data" . $count] = $item;
            }
            $count++;
        }
    } else {
// get the default page of module.
        switch (ARCDEFAULTTYPE) {
            case "module":
                arcSetPage(ARCDEFAULTPAGE, null);
                break;
            case "page":
                arcSetPage("page", ARCDEFAULTPAGE);
                break;
            default:
                arcSetPage("error", "404");
                break;
        }
    }
}

// get header
function arcGetHeader() {
    arcSplitURL();

    $page = Page::getBySEOURL(arcGetURLData("data1"));
    echo "<title>";
    if (!empty($page->title)) {
        echo $page->title;
    } else {
        echo ARCTITLE;
    }

    echo "</title>" . PHP_EOL;
    echo "\t<meta name='author' content='" . ARCAUTHOR . "'>" . PHP_EOL;
    if (!empty($page->metadescription)) {
        echo "\t<meta name='description' content='" . $page->metadescription . "'>" . PHP_EOL;
    }
    if (!empty($page->metakeywords)) {
        echo "\t<meta name='keywords' content='" . $page->metakeywords . "'>" . PHP_EOL;
    }
    if (!empty($page->title)) {
        echo "\t<link rel='alternate' href='http://" . $_SERVER['HTTP_HOST'] . ARCWWW . "page/" . $page->seourl . "'>" . PHP_EOL;
    }
    echo "\t<link rel='icon' href='" . ARCFAVICON . "'>" . PHP_EOL;

    arcGetCSSJavascript("css", "css/");
    if (!empty(arcGetURLData("module"))) {
        arcGetCSSJavascript("css", "modules/" . arcGetURLData("module") . "/css/");
    }
    
    arcGetTheme();

    arcGetCSSJavascript("js", "js/");
    if (!empty(arcGetURLData("module"))) {
        arcGetCSSJavascript("js", "modules/" . arcGetURLData("module") . "/js/");
    }
}

// output css and javascript
function arcGetCSSJavascript($type, $path) {
    if (file_exists(arcGetPath(true) . $path)) {
        $files = scandir(arcGetPath(true) . $path);
        foreach ($files as $file) {
            if ($file != "." && $file != ".." && !is_dir(arcGetPath(true) . $path . $file)) {
                if ($type == "js") {
                    echo "\t<script src='" . arcGetPath() . $path . $file . "'></script>" . PHP_EOL;
                } elseif ($type == "css") {
                    echo "\t<link href='" . arcGetPath() . $path . $file . "' rel='stylesheet'>" . PHP_EOL;
                }
            }
        }
    }
}

// set page
function arcSetPage($name, $data = null) {
    $GLOBALS["arc_url_data"] = array();
    $GLOBALS["arc_url_data"]["module"] = $name;
    $count = 1;
    $url = explode("/", $data);
    foreach ($url as $item) {
        $GLOBALS["arc_url_data"]["data" . $count] = $item;
        $count++;
    }
}

// get content
function arcGetContent() {
    if (arcGetURLData("module") == "logout") {
        session_unset();
        session_destroy();
        arcRedirect();
        return;
    }

    $timeout = ARCSESSIONTIMEOUT * 60;
    if (isset($_SESSION["LAST_ACTIVITY"]) && (time() - $_SESSION["LAST_ACTIVITY"] > $timeout)) {
        session_unset();
        session_destroy();
        arcSetPage("error", "419");
    } else {
        $_SESSION["LAST_ACTIVITY"] = time(); // update last activity time stamp   
        
        $page = Page::getBySEOURL(arcGetURLData("module"));
        if ($page->id != 0) {
            // if we have a page set it.
            arcSetPage("page", arcGetURLData("module"));
        }

        if (!file_exists(arcGetPath(true) . "modules/" . arcGetURLData("module"))) {
            // module not found.
            arcSetPage("error", "404");
        } else {

            $permissions;
            if (!empty(arcGetUser())) {
                $user = arcGetUser();
                $group = $user->getGroup();
                $permissions = $group->getPermissions();
            } else {
                $group = UserGroup::getByName("Anyone");
                $permissions = $group->getPermissions();
            }

            $pCheck = arcGetURLData("module") . "/" . arcGetURLData("data1");

            if (arcGetURLData("module") != "page") {
                $pCheck = "module/" . arcGetURLData("module");
            }

            if (!UserPermission::hasPermission($permissions, $pCheck)) {
                arcSetPage("error", "403");
            }
        }
    }

    if (arcGetURLData("data1") == "administration") {
        include_once arcGetPath(true) . "modules/" . arcGetURLData("module") . "/administration/index.php";
    } else {
        include_once arcGetPath(true) . "modules/" . arcGetURLData("module") . "/index.php";
    }
}

// get user
function arcGetUser() {
    if (isset($_SESSION["arc_user"])) {
        return unserialize($_SESSION["arc_user"]);
    }
    return null;
}

// set user
function arcSetUser($user) {
    $_SESSION["arc_user"] = serialize($user);
}

// redirect to default
function arcRedirect($destination = null) {
    echo "<script>window.location = '";
    if (empty($destination)) {
        echo ARCWWW;
    } else {
        echo $destination;
    }
    echo "';</script>";
}

// get dispatch url
function arcGetDispatch() {
    if (arcGetURLData("data1") == "administration") {
        echo ARCWWW . "modules/" . arcGetURLData("module") . "/administration/dispatch.php";
    } else {
        echo ARCWWW . "modules/" . arcGetURLData('module') . "/dispatch.php";
    }
}

// get module root
function arcGetModulePath($filesystem = false) {
    if ($filesystem == true) {
        if (arcGetURLData("data1") == "administration") {
            return arcGetPath(true) . "modules/" . arcGetURLData("module") . "/administration/";
        } else {
            return arcGetPath(true) . "modules/" . arcGetURLData("module");
        }
    }
    if (arcGetURLData("data1") == "administration") {
        return ARCWWW . arcGetURLData("module") . "/administration/";
    } else {
        return ARCWWW . arcGetURLData("module");
    }
}

// get menu
function arcGetMenu() {
    $modules = scandir(arcGetPath(true) . "modules");
    $module_list = array();
    $groups = array();

    $group = new UserGroup();
    if (!empty(arcGetUser())) {
        $user = arcGetUser();
        $group = $user->getGroup();
    } else {
        $group->getByID("3");
    }

    $permissions = $group->getPermissions();
    $perms = new UserPermission();

    foreach ($modules as $module) {
        $module_info = "";
        if ($module != ".." && $module != ".") {
            // module menu
            if (file_exists(arcGetPath(true) . "modules/" . $module . "/info.php")) {
                if ($perms->hasPermission($permissions, "module/" . $module)) {
                    include arcGetPath(true) . "modules/" . $module . "/info.php";
                    if (isset($module_info["menu"]) && count($module_info["menu"] > 0)) {
                        foreach ($module_info['menu'] as $menu) {
                            $menu["module"] = $module;
                            if (empty($menu["group"])) {
                                $module_list[] = $menu;
                            } else {
                                $groups[$menu["group"]][] = $menu;
                            }
                        }
                    }
                }
            }

            // module administration menu
            if ($group->name == "Administrators") {
                if (file_exists(arcGetPath(true) . "modules/" . $module . "/administration/info.php")) {
                    include arcGetPath(true) . "modules/" . $module . "/administration/info.php";
                    if (isset($module_info["menu"]) && count($module_info["menu"] > 0)) {
                        foreach ($module_info["menu"] as $menu) {
                            $menu["module"] = $module;
                            $menu["group"] = "Administration";
                            $groups["Administration"][] = $menu;
                        }
                    }
                }
            }
        }
    }

// logout menu (last item)
    $module_info['name'] = "Logout";
    $module_info["icon"] = "fa-lock";
    $module_info["divider"] = false;
    $module_info["group"] = "";
    $module_info["module"] = "logout";
    $module_list[] = $module_info;

    echo "\t<ul class='nav navbar-nav navbar-right'>" . PHP_EOL;

    if (count($groups) > 0) {
        foreach ($groups as $key => $value) {
            echo "\t\t<li class='dropdown'>";
            echo "<a href='#' class='dropdown-toggle' data-toggle='dropdown'>";
            echo "<span class='fa ";
            if ($key == "Administration") {
                echo "fa-cogs";
            } else {
                echo "fa-list";
            }
            echo "'></span>&nbsp;" . $key . " <span class='caret'></span></a>";
            echo "\t\t<ul class='dropdown-menu' role='menu'>";
            arcProcessMenuItems($value);
            echo "</ul></li>" . PHP_EOL;
        }
    }

    arcProcessMenuItems($module_list);

    echo "</ul>" . PHP_EOL;
}

// process menu items
function arcProcessMenuItems($items) {
    foreach ($items as $item) {
        if (isset($item["divider"]) && $item["divider"] == true) {
            echo "\t\t\t<li class='divider'></li>" . PHP_EOL;
        }
        echo "\t\t\t<li><a href='" . arcGetPath() . $item["module"];
        if ($item["group"] == "Administration") {
            echo "/administration";
        } elseif (isset($item["url"]) && !empty($item["url"])) {
            echo $item["url"];
        }
        echo "'><span class='fa " . $item["icon"] . "'></span> " . $item["name"] . "</a></li>" . PHP_EOL;
    }
}

// get modules
function arcGetModules() {
    $modules = scandir(arcGetPath(true) . "modules");
    $module_list = array();
    foreach ($modules as $module) {
        $module_info["name"] = 'Unknown';
        $module_info["description"] = 'No description provided';
        $module_info["version"] = '0.0.0.0';
        $module_info["author"] = 'Unknown';
        $module_info["email"] = 'Unknown';
        $module_info["www"] = 'Unknown';
        $module_info["system"] = false;
        if ($module != ".." && $module != ".") {
            if (file_exists(arcGetPath(true) . "modules/" . $module . "/info.php")) {
                include arcGetPath(true) . "modules/" . $module . "/info.php";

                $module_info["module"] = $module;
                $module_list[] = $module_info;
            } elseif (file_exists(arcGetPath(true) . "modules/" . $module . "/administration/info.php")) {
                include arcGetPath(true) . "modules/" . $module . "/administration/info.php";

                $module_info["module"] = $module;
                $module_list[] = $module_info;
            }
        }
    }
    return $module_list;
}

//get theme
function arcGetTheme() {
    if (!empty(arcGetUser())) {
        $user = arcGetUser();
        $theme = $user->getSettingByKey("ARC_THEME");
        if (!empty($theme->setting)) {
            echo "\t<link href='" . arcGetPath() . "css/themes/" . $theme->setting . ".min.css' rel='stylesheet'>" . PHP_EOL;
            return;
        }
    }
    echo "\t<link href='" . arcGetPath() . "css/themes/" . ARCDEFAULTTHEME . ".min.css' rel='stylesheet'>" . PHP_EOL;
}
